presto-hive/src/main/java/com/facebook/presto/hive/HiveSplitSource.java

```diff
@@ -104,7 +127,106 @@ void fail(Throwable e)
+                        if (e.getErrorCode().equals(NOT_SUPPORTED)) {
+                            // TODO: log information about not supported type
+                            continue;
+                        }
+
+                        throw e;
+                    }
+                    if (predicateDomain.overlaps(Domain.singleValue(type, objectToWrite))) {
+                        result.add(new HiveSplit(
+                                hiveSplit.getClientId(),
+                                hiveSplit.getDatabase(),
+                                hiveSplit.getTable(),
+                                hiveSplit.getPartitionName(),
+                                hiveSplit.getPath(),
+                                hiveSplit.getStart(),
+                                hiveSplit.getLength(),
+                                hiveSplit.getFileSize(),
+                                hiveSplit.getSchema(),
+                                hiveSplit.getPartitionKeys(),
+                                hiveSplit.getAddresses(),
+                                hiveSplit.getBucketNumber(),
+                                hiveSplit.isForceLocalScheduling(),
+                                hiveSplit.getEffectivePredicate().intersect(runtimeTupleDomain),
+                                hiveSplit.getColumnCoercions()));
+                        break;
+                    }
+                }
+            }
+        }
+        return result.build();
+    }
+
+    class DomainsCache
+    {
+        private final Map<HiveColumnHandle, Domain> domains;
+        private final Map<String, Collection<Domain>> domainsCache = new HashMap<>();
+
+        public DomainsCache(Map<HiveColumnHandle, Domain> domains)
+        {
+            this.domains = ImmutableMap.copyOf(domains);
+        }
+
+        public Collection<Domain> getDomains(String partitionName)
+        {
+            String partitionNameLower = partitionName.toLowerCase(Locale.ENGLISH);
+            Collection<Domain> relevantDomains = domainsCache.get(partitionNameLower);
+            if (relevantDomains == null) {
+                relevantDomains = domains.entrySet().stream()
+                        .filter(entry -> entry.getKey().getName().equalsIgnoreCase(partitionNameLower))
+                        .map(Map.Entry::getValue)
+                        .collect(toImmutableList());
+                domainsCache.put(partitionNameLower, relevantDomains);
+            }
+
+            return relevantDomains;
+        }
+    }
+
+    private List<DynamicFilterDescription> getAvailableDynamicFilterDescriptions()
+    {
+        ImmutableList.Builder<DynamicFilterDescription> dynamicFilterDescriptionBuilder = ImmutableList.builder();
+        for (Future<DynamicFilterDescription> descriptionFuture : filters) {
+            Optional<DynamicFilterDescription> description = MoreFutures.tryGetFutureValue(descriptionFuture);
```

Comment:
@maciejgrzybek 
While testing I figured out that this would always return empty Optional. If we just change it to `descriptionFuture.get()` query will get stuck forever. Seems like some issue. Something seems to be off here.